Responds to simultaneous requests

indykid opened the pull request:
- Server receives the client connection
- Server instantiates Runnable Worker with the connections and response controller
- Server gets ExecutorService to start a separate thread
- Worker is run on the thread

Commit:
Implements Runnable Worker class

Commit:
Responds to simultaneous requests

- makes use of the Worker class implemented earlier
- relies on ExecutorService for managing threads

Commit:
Isolates knowledge about the port to one place

- also improves coverage for PublicResourceResponder

mateuadsuara commented on src/main/java/kg/jarkyn/cobspecserver/Server.java:
```diff
@@ -28,6 +32,6 @@ private ClientSocket nextClient() {
 
     private void respondTo(ClientSocket client) {
-        responseController.respond(client);
+        executor.execute(new Worker(client, responseController));
```
I believe `executor.execute(new Worker(client, anotherResponseController));` will not make your tests fail. 

Maybe `ExecutorServiceDouble. getRunnable().run()` might help checking that. Not sure how valuable that test would be though...

indykid replied:
hmm i see, i will look into it, thank you!

mateuadsuara commented on src/main/java/kg/jarkyn/cobspecserver/Runner.java:
```diff
@@ -12,55 +12,64 @@
+    private static final int DEFAULT_PORT = 5000;
+    private static final boolean KEEP_RUNNING = true;
+    private static final String DOMAIN = "http://localhost";
+    private static final String PUBLIC_RESOURCE_PATH = "/Users/Jarkyn/Projects/8thLight/JAVA/cob_spec/public";
```
I don't think `DEFAULT_PORT` and `PUBLIC_RESOURCE_PATH` should be constants. Those values happen to be the same CobSpec uses by default.
But https://github.com/8thlight/cob_spec#http-server specifies that CobSpec expects to be able to send them as arguments. 
I don't know if this is in the scope of your iteration, I would ask your mentors if you are not sure.

indykid replied:
:) someone else suggested that they should be, they weren't before.
It was not a requirement for me to parse args, I have a very clear idea how I would do it when given the time.

mateuadsuara commented on src/main/java/kg/jarkyn/cobspecserver/Runner.java:
```diff
@@ -12,55 +12,64 @@
 
     private static ListenerSocket listener() {
-        return new ListenerSocket(getServerSocket(), KEEP_RUNNING);
+        return new ListenerSocket(setupServerSocket(), KEEP_RUNNING);
```
Why should `true` always be sent to `ListenerSocket`? 
Is there a case where you want to create a `ListenerSocket` that is never going to accept anything (passing `false`)?
What about not passing that argument and initializing the `accepting` instance variable to `true` in the constructor?

I guess the purpose is that in some tests you want to use a ListenerSocket that finished accepting connections. What about the following for those cases?

```
listenerSocket = new ListenerSocket(serverSocket)
listenerSocket.stop()
```

indykid replied:
What you are suggesting was my initial implementation, and then I had feedback from two people that I should pass the flag in. I personally prefer what you are suggesting, but I obliged with the feedback received :(
I may put this back to how I had it originally, to me this seems quite ugly

mateuadsuara commented on src/test/java/kg/jarkyn/cobspecserver/WorkerTest.java:
```diff
@@ -0,0 +1,23 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class WorkerTest {
```
What should happen when exceptions on the socket communication occur? At the moment, `RequestParser` and `StreamHandler` just rethrow them wrapped in `RuntimeException`...
Can they occur?

Have you tried loading the server in your browser and hold cmd+r (reload) in one of the pages?
Have you tried putting the server under stress (several concurrent requests)?

There are tools to help with that: https://en.wikipedia.org/wiki/Web_server_benchmarking

indykid replied:
I have looked into when IOException can be thrown and it seemed the cases in which it may occur it is justifiable to have a RuntimeException. Ideally there could be something more elegant, a combination of logging for developer and an error message to the user, however this is a little beyond my stories.

mateuadsuara commented on src/test/java/kg/jarkyn/cobspecserver/ServerTest.java:
```diff
@@ -1,43 +1,45 @@
     }
 
     @Test
     public void runsUntilStopped() {
```
I first thought: _when is the server stopped?_. Then I looked into `ListenerDouble` because that magic `2` seemed to be related.
It is not clear that `ListenerDouble` stops after two connections. 

But I wonder: Is `ListenerSocket` the appropriate place to stop the server? When should the server stop? why?
The only way to make `ListenerSocket` stop is by calling its `stop()` method. How should one stop the server?

```
ListenerSocket listener = createListener();
Server server = new Server(listener, controller);
server.run();
listener.stop();
```

Seems like the server is not encapsulating its internals: we still need to know its content to be able to do things with it.
What about making the `Server` able to stop?

```
Server server = new Server(createListener(), controller);
server.run();
server.stop();
```

On the other hand, I think the call to `run` is a blocking call, so this might not be possible with that. But then, that goes back to the question: _How should one stop the server?_ Killing the process?

I am not sure how complicated to implement stopping the server would be. And I don't think is valuable enough to be in the scope of your iteration. But it is interesting to think about it.

indykid replied:
wold love to hear more about this, along with other points you have made

mateuadsuara commented on src/main/java/kg/jarkyn/cobspecserver/Server.java:
```diff
@@ -4,11 +4,15 @@
     private ResponseController responseController;
 
-    public Server(ListenerSocket listener, ResponseController responseController) {
+    public Server(ExecutorService executor, ListenerSocket listener, ResponseController responseController) {
```
Creating dependencies to concrete classes is a typical smell. There is a missing opportunity to unveil an interesting abstraction.

Why to put the multithreading code here? it seems like `ResponseController` knows how to respond to a `client`. It could delegate to workers as well.

But if we do that, we might put too much responsibilities on that concrete class. This might give you a hint that an abstraction is wanting to appear.
That could guide you to a design pattern (composite probably).

Not sure you have space in your iteration for that refactoring though...

indykid replied:
would be good to talk about it more! thank you, and yes, not sure I actually have enough time for this before today.

ukutaht commented on src/test/java/kg/jarkyn/cobspecserver/doubles/ExecutorServiceDouble.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.List;
+import java.util.concurrent.*;
+
+public class ExecutorServiceDouble implements ExecutorService {
```
Is there a class that you could subclass instead? The interface is forcing you to implement a lot of methods you don't care about, I'd try subclassing and overriding just the important methods for your tests

ukutaht commented on src/test/java/kg/jarkyn/cobspecserver/responders/PublicResourceResponderTest.java:
```diff
@@ -44,6 +44,15 @@ public void respondsWithDirectoryListing() {
     }
 
+    @Test
+    public void respondsWithNotFound() {
```
I find these tests too low-level. Overall the intention is clear: when a url that does not map to a resource is hit, 404 is returned. That makes sense, however this test is matching the _exact_ http response that is returned. Seeing the HTTP version and carriage returns in the assertions is distracting and irrelevant to what's actually tested.

The assertion of this test could read like this:

``` java
 assertEquals("404", response.getStatus());
```

Or, with an enum:

``` java
 assertEquals(Status.NOT_FOUND, response.getStatus());
```

indykid replied:
ok, understood

Commit:
Changes Response signature to use Status enum

- moves status line formatting responsibility to Response
- helps remove knowledge about format of the status from tests

Commit:
Changes Response signature to take headers as map

- moves responsibility for header formatting from responders to Response
- removes knowledge about header format from tests

Commit:
Simplifies Server testing
